src/formpack/utils/kobo_locking.py

```diff
@@ -0,0 +1,177 @@
+# coding: utf-8
+from __future__ import (unicode_literals, print_function,
+                        absolute_import, division)
+
+import io
+import itertools
+from collections import OrderedDict
+
+from .xls_to_ss_structure import xls_to_dicts
+from formpack.constants import (
+    KOBO_LOCKING_RESTRICTIONS,
+    KOBO_LOCK_COLUMN,
+    KOBO_LOCK_KEY,
+    KOBO_LOCK_SHEET,
+)
+from formpack.utils.exceptions import FormPackLibraryLockingError
+
+def get_kobo_locking_profiles(xls_file_object: io.BytesIO) -> list:
+    """
+    Return the locking profiles, if there are any, in a dictionary structure
+    from an XLSForm matrix. For example, the following matrix structure:
+
+    # kobo--locking-profiles
+    |    restriction    | profile_1 | profile_2 |
+    |-------------------|-----------|-----------|
+    | choice_add        | locked    |           |
+    | choice_delete     |           | locked    |
+    | choice_label_edit | locked    |           |
+    | choice_order_edit | locked    | locked    |
+
+    Will be transformed into the following JSON structure:
+    [
+        {
+            "name": "profile_1",
+            "restrictions": [
+                "choice_add",
+                "choice_label_edit",
+                "choice_order_edit"
+            ],
+        },
+        {
+            "name": "profile_2",
+            "restrictions": [
+                "choice_delete",
+                "choice_order_edit"
+            ],
+        }
+    ]
+    """
+    survey_dict = xls_to_dicts(xls_file_object)
+
+    if KOBO_LOCK_SHEET not in survey_dict:
+        return
+
+    locks = survey_dict[KOBO_LOCK_SHEET]
+    # Get a unique list of profile names if they have at least one value set to
+    # 'locked' from the matrix of values
+    profiles = set(itertools.chain(*[lock.keys() for lock in locks]))
```

Comment:
this is clever! but i wonder if it's worth the extra complexity over a simpler construct like
```python
profiles = set()
for lock in locks:
    profiles.update(lock.keys())
```
(i think this is also simpler for the machine to execute)

also, does this step really care about the value `locked` as the comment indicates? if the comment is meant to apply to a larger section of code, maybe some additional whitespace would help.

Reply:
That's a good point 👌 definitely easier to read too.

You're right, I think that check actually happens in the `xls_to_lists()` method that `xls_to_dicts()` calls.
In essence, if the xlsform has something like this:

**`kobo–locking-profiles`**
restriction | core | flex | foo
--- | --- | --- | ---
choice_add | locked | locked

Then the locking profiles will only include `core` and `flex` because they have at least one restriction set.